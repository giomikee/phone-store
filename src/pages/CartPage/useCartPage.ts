import { useDispatch } from 'react-redux';
import type { AddedPhone } from '../../interfaces';
import { useState } from 'react';
import { removeFromCart } from '../../state/cart/cart.slice';

export function useCartPage() {
    const dispatch = useDispatch();

    const [phoneToRemove, setPhoneToRemove] = useState<AddedPhone | null>(null);
    const [cartIndexToRemove, setCartIndexToRemove] = useState<number | null>(null);
    const [isDialogOpen, setIsDialogOpen] = useState<boolean>(false);

    const handleClickRemove = (phone: AddedPhone, index: number) => {
        setPhoneToRemove(phone);
        setCartIndexToRemove(index);
        setIsDialogOpen(true);
    };

    const handleCloseDialog = (isDeleteConfirmed: boolean) => {
        if (cartIndexToRemove === null) {
            return;
        }

        if (isDeleteConfirmed) {
            dispatch(removeFromCart(cartIndexToRemove));
        }

        setPhoneToRemove(null);
        setCartIndexToRemove(null);
        setIsDialogOpen(false);
    };

    return {
        phoneToRemove,
        isDialogOpen,
        cartIndexToRemove,
        setPhoneToRemove,
        setCartIndexToRemove,
        handleClickRemove,
        handleCloseDialog,
    };
}