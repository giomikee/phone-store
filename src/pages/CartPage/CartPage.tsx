import { Container, Divider, Typography } from '@mui/material';
import { HEADINGS } from '../../constants';
import { getCartTotalPrice } from '../../utils/getCartTotalPrice';
import CartActions from '../../components/CartActions/CartActions';
import CartItems from '../../components/CartItems/CartItems';
import ConfirmDeleteDialog from '../../components/ConfirmDeleteDialog/ConfirmDeleteDialog';
import { useCartPage } from './useCartPage';
import { useStoreSelectors } from '../../state/useStoreSelectors';

const CartPage = () => {
    const { cartPhones } = useStoreSelectors();
    const { isDialogOpen, phoneToRemove, handleClickRemove, handleCloseDialog }= useCartPage();

    return (
        <Container>
            <Typography
                component={HEADINGS.h2}
                variant={HEADINGS.h5}
            >
                Carrito de compra
            </Typography>

            <Divider sx={{ my: 2 }} />

            <CartItems
                onClickRemove={handleClickRemove}
                phones={cartPhones}
            />

            <Divider sx={{ my: 2 }} />

            <CartActions
                totalPrice={getCartTotalPrice(cartPhones)}
                isBuyDisabled={!cartPhones.length}
            />

            {
                isDialogOpen && phoneToRemove &&
                    <ConfirmDeleteDialog
                        onClose={handleCloseDialog}
                        phone={phoneToRemove}
                    />
            }

        </Container>
    )
}

export default CartPage