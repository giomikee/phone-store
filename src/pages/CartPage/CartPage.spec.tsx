import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import CartPage from './CartPage';
import { BrowserRouter } from 'react-router-dom';


vi.mock('react-redux', async () => ({
    ...(await vi.importActual('react-redux')),
    useDispatch: vi.fn().mockReturnValue(vi.fn()),
}));
vi.mock('react-router-dom', async () => ({
    ...(await vi.importActual('react-router-dom')),
    useNavigate: vi.fn().mockReturnValue(vi.fn()),
}));

vi.mock('../../state/cart/cart.slice');

vi.mock('./useCartPage', () => ({
    useCartPage: vi.fn().mockReturnValue({
        isDialogOpen: false,
        phoneToRemove: {
            id: 'id',
            name: 'name',
            brand: 'brand',
            basePrice: 0,
            selectedColor: {
                hexCode: 'hexCode',
                name: 'name',
                imageUrl: 'imageUrl'
            },
            selectedStorage: {
                capacity: 'capacity',
                price: 3.1
            },
        },
        handleClickRemove: vi.fn(),
        handleCloseDialog: vi.fn(),
    }),
}));

vi.mock('../../state/useStoreSelectors');

describe('Given CartPage component', () => {
    describe('When the component is rendered', () => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<CartPage />, { wrapper: BrowserRouter });
            expect(container).toMatchSnapshot();
        });
    })
})