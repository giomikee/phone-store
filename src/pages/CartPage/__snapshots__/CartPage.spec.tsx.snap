// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given CartPage component > When the component is rendered > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
  >
    <h2
      class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
    >
      Carrito de compra
    </h2>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
    />
    <p
      class="MuiTypography-root MuiTypography-body1 css-1xrrpwr-MuiTypography-root"
    >
      No hay móviles en el carrito.
    </p>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
    />
    <div
      class="MuiStack-root css-1cic80u-MuiStack-root"
    >
      <a
        href="/"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary css-efdih2-MuiButtonBase-root-MuiButton-root"
          data-testid="continue-shopping-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="SmartphoneIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 1.01 7 1c-1.1 0-2 .9-2 2v18c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2V3c0-1.1-.9-1.99-2-1.99M17 19H7V5h10z"
              />
            </svg>
          </span>
          Continuar comprando
        </button>
      </a>
      <div
        class="MuiStack-root css-1g8n1bs-MuiStack-root"
      >
        <h3
          class="MuiTypography-root MuiTypography-h5 css-1674ucw-MuiTypography-root"
        >
          Total: 
          0.00 €
        </h3>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSuccess Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSuccess css-1cobkga-MuiButtonBase-root-MuiButton-root"
          data-testid="buy-button"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="CreditScoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 4H4c-1.11 0-1.99.89-1.99 2L2 18c0 1.11.89 2 2 2h5v-2H4v-6h18V6c0-1.11-.89-2-2-2m0 4H4V6h16zm-5.07 11.17-2.83-2.83-1.41 1.41L14.93 22 22 14.93l-1.41-1.41z"
              />
            </svg>
          </span>
          <span
            class="MuiButton-loadingWrapper"
            style="display: contents;"
          />
          Comprar productos
        </button>
      </div>
    </div>
  </div>
</div>
`;
